protos/dns/dns.go

```diff
@@ -0,0 +1,712 @@
+	}
+	trans.Dst = common.Endpoint{
+		Ip:   tuple.Dst_ip.String(),
+		Port: tuple.Dst_port,
+		Proc: string(cmd.Dst),
+	}
+	return trans
+}
+
+type Dns struct {
+	// Configuration data.
+	Ports               []int
+	Send_request        bool
+	Send_response       bool
+	Include_authorities bool
+	Include_additionals bool
+
+	// Map of active DNS transactions. The map key is the HashableDnsTuple
+	// associated with the request. Use the put, lookup, and deleteTransaction
+	// methods to make map access concurrency-safe.
+	transactionsMap   map[HashableDnsTuple]*DnsTransaction
+	transactionsMutex sync.Mutex
+
+	results chan common.MapStr // Channel where results are pushed.
+}
+
+// putTransaction puts a transaction into the transaction map. If the
+// key already exists then the exiting entry will be overridden. The
+// key should be the HashableDnsTuple associated with the request (src
+// is the requestor).
+func (dns *Dns) putTransaction(h HashableDnsTuple, trans *DnsTransaction) {
+	dns.transactionsMutex.Lock()
+	defer dns.transactionsMutex.Unlock()
+	dns.transactionsMap[h] = trans
+}
+
+// lookupTransaction returns the transaction associated with the given
+// HashableDnsTuple. The lookup key should be the HashableDnsTuple associated
+// with the request (src is the requestor). Nil is returned if the entry
+// does not exist.
+func (dns *Dns) lookupTransaction(h HashableDnsTuple) *DnsTransaction {
+	dns.transactionsMutex.Lock()
+	defer dns.transactionsMutex.Unlock()
+	return dns.transactionsMap[h]
+}
+
+// deleteTransaction deletes an entry from the transaction map and returns
+// the deleted element. If the key does not exist then nil is returned.
+func (dns *Dns) deleteTransaction(h HashableDnsTuple) *DnsTransaction {
+	dns.transactionsMutex.Lock()
+	defer dns.transactionsMutex.Unlock()
+	t := dns.transactionsMap[h]
+	delete(dns.transactionsMap, h)
+	return t
+}
+
+func (dns *Dns) initDefaults() {
+	dns.Send_request = false
+	dns.Send_response = false
+	dns.Include_authorities = false
+	dns.Include_additionals = false
+}
+
+func (dns *Dns) setFromConfig(config config.Dns) error {
+
+	dns.Ports = config.Ports
+
+	if config.Send_request != nil {
+		dns.Send_request = *config.Send_request
+	}
+	if config.Send_response != nil {
+		dns.Send_response = *config.Send_response
+	}
+	if config.Include_authorities != nil {
+		dns.Include_authorities = *config.Include_authorities
+	}
+	if config.Include_additionals != nil {
+		dns.Include_additionals = *config.Include_additionals
+	}
+	return nil
+}
+
+func (dns *Dns) Init(test_mode bool, results chan common.MapStr) error {
+	dns.initDefaults()
+	if !test_mode {
+		dns.setFromConfig(config.ConfigSingleton.Protocols.Dns)
+	}
+
+	dns.transactionsMap = make(map[HashableDnsTuple]*DnsTransaction, TransactionsHashSize)
+	dns.transactionsMutex = sync.Mutex{}
+
+	dns.results = results
+
+	return nil
+}
+
+func (dns *Dns) GetPorts() []int {
+	return dns.Ports
+}
+
+func (dns *Dns) ParseUdp(pkt *protos.Packet) {
```

Comment:
+1, this transforms a crash into a log entry saying: "There was an issue but we recovered, please report: <<stacktrace>>"